### ISSUE TITLE
[Bug][AMD] Custom AR graph registration corrupts DSA DP-attention graph-skew replay

### ISSUE THREAD

**Issue by tanth47** [sgl-project/sglang#36071]
## Summary

On ROCm, custom all-reduce direct graph-input registration is unsafe when DSA
DP-attention lanes independently replay different decode graphs. On GLM-5.2
TP8/DP4 this can make the collective incomplete, turn every output logit into
NaN, and emit token ID 0 repeatedly (`!!!!!!!!`).

The failure does not require MTP, overlap scheduling, AITER prefill, or AITER
all-reduce fusion. Eager decode, plain NCCL, and the custom-AR stable copy-in
path are clean under the same model and parallelism.

## Environment

- 8 x AMD MI350X (`gfx950`)
- ROCm 7.2.4
- SGLang revision `db570fe619ae15d8bd4ee04f146bc8234c52a20d`
- AITER revision `c16d44b93a528b2a4bfd6d8d3409116d465872a9`
- Model: `amd/GLM-5.2-MXFP4`, real weights
- KV cache: BF16
- TP8, DP4 with DP attention, EP8
- TileLang DSA prefill and decode

## Minimal server configuration

```bash
SGLANG_USE_AITER=1 \
SGLANG_USE_ROCM700A=1 \
SGLANG_ROCM_FUSED_DECODE_MLA=1 \
python3 -u -m sglang.launch_server \
  --model-path amd/GLM-5.2-MXFP4 \
  --served-model-name glm-5.2 \
  --tp-size 8 \
  --dp-size 4 \
  --enable-dp-attention \
  --ep-size 8 \
  --moe-a2a-backend none \
  --kv-cache-dtype bf16 \
  --mem-fraction-static 0.85 \
  --attention-backend dsa \
  --dsa-prefill-backend tilelang \
  --dsa-decode-backend tilelang \
  --dsa-topk-backend sgl-kernel \
  --enable-aiter-allreduce-fusion \
  --disable-overlap-schedule \
  --cuda-graph-backend-decode full \
  --cuda-graph-max-bs-decode 128 \
  --cuda-graph-bs-decode 1 2 4 8 16 32 64 128 \
  --cuda-graph-backend-prefill disabled \
  --quantization quark \
  --trust-remote-code
```

MTP is intentionally disabled in the focused reproducer.

## Actual result

Decode eventually emits token ID 0 with null logprobs:

```text
We need to write a correct Python function!!!!!!!!

output_ids:
[1654, 1184, 311, 3270, 264, 4396, 13020, 729, 0, 0, 0, 0, 0, 0, 0, 0]
```

Both attention-TP ranks in the routed DP lane report all-NaN logits:

```text
shape=(1, 154880) finite=0 nan=154880
```

## Root cause

This is not a DSA attention-kernel error or a shared graph-memory-pool pointer
collision.

During capture, AITER/SGLang direct custom-AR registration associates every
captured call with a `RankData` slot. Post-capture registration fills that slot
with peer pointers from the same captured-call ordinal. Replay has no runtime
graph-id handshake and therefore assumes every communicator rank is replaying
the same captured graph/call ordinal.

DSA DP-attention lanes can select different graphs concurrently. A TP2 pair
replaying graph 0 consequently reads graph-0 buffers on peers currently
producing graph 1, 2, or 3. Those buffers are stale/inactive and the global
collective is incomplete.

## Minimal causal proof

A standalone eight-GPU test captures four same-shape BF16 graphs with unique
per-graph values:

| Input path | Replay pattern | Failed rank-results |
|---|---|---:|
| direct registration | all ranks replay the same graph | 0/32 |
| direct registration | DP4/TP2 graph skew | 8/8 |
| stable copy-in | identical graph skew | 0/8 |

Expected global sum: `2080` on every rank. Direct/skew mode returns `206`,
`414`, `624`, and `832`, exactly the four active TP2-pair contributions.
Copy-in returns `2080` on all eight ranks.

Instrumentation of 126,416 capture/registration events found no registration
count, IPC-handle/offset, or capture-to-registration drift. A stronger
release/acquire system-scope barrier still failed 8/8 full-model requests.

## Full-model isolation

| Collective path | OSL32 result |
|---|---:|
| AITER custom AR, direct graph registration | 4/4 corrupt |
| SGLang custom AR, direct graph registration | 4/4 corrupt |
| Plain NCCL | 0/4 corrupt |
| AITER custom AR, stable copy-in | 0/4 corrupt |

Copy-in subsequently passed 32/32 sequential and 32/32 concurrent requests,
including the production MTP + overlap configuration.

## Fix

#36088 retains custom AR but uses its stable pre-registered copy-in buffer
during graph capture for `DSA + DP attention`. This removes graph identity from
the peer-pointer contract. The measured decode throughput delta was -0.40%.

A future zero-copy design would need runtime active-graph-slot indirection or
allocator integration that makes every graph produce into the same stable
registered input address.

**Comment by tanth47** [comment]
Root cause update: this is not an AITER-prefill or DSA-kernel failure. It is
the custom-all-reduce direct graph-input registration path interacting with the
DSA dense/sparse dual decode graphs under DP attention.

The first eight generated tokens can be coherent, so the corrected minimal
gate uses sequential OSL32 (the earlier OSL8 gate was a false negative).

Same TP8/DP4/EP8 config, no MTP, full decode graphs:

| Collective path | OSL32 result |
|---|---:|
| AITER custom AR, fusion on | 4/4 corrupt |
| AITER custom AR, fusion off | 4/4 corrupt |
| SGLang custom AR | 4/4 corrupt |
| Plain NCCL | 0/4 corrupt |
| AITER custom AR with graph capture forced through its pre-registered copy-in buffer | 0/4 corrupt |

The copy-in result also passed 32/32 sequential requests and 32/32 concurrent
c8 x 4 OSL64 requests. With the full target setup (AITER prefill, TileLang
decode, MTP, overlap scheduling), it again passed 32/32 sequential and 32/32
concurrent requests.

Both custom-AR implementations normally bind graph-owned input pointers
directly during capture. DSA DP attention captures dense and sparse variants
for each bucket in a shared graph pool. The evidence is consistent with an
unsafe peer mapping when a local pool address is reused across variants with a
different cross-rank allocation layout. Staging through the communicator's
stable, pre-registered input buffer removes that dependency and fixes the
corruption.

The focused fix is in #36088. It enables copy-in capture for DSA DP attention
for both AITER and legacy SGLang custom AR, while retaining direct registration
elsewhere.

**Comment by tanth47** [comment]
Root cause correction and causal proof are now available in #36088.

The failure is **not** caused by dense DSA attention itself or by two graph
memory pools colliding. Direct custom-AR graph registration binds every
captured call to peer pointers from the same graph/call ordinal. DSA DP lanes
may replay different graphs concurrently, but the custom-AR kernel has no
runtime graph-id handshake. A lane therefore reads stale/inactive graph-local
buffers on ranks replaying another graph, producing an incomplete all-reduce.

Standalone 8-GPU proof using four same-shape BF16 graphs:

| Mode | Replay pattern | Failed rank-results |
|---|---|---:|
| direct registration | lockstep graph selection | 0/32 |
| direct registration | DP4/TP2 graph skew | 8/8 |
| stable copy-in buffer | identical graph skew | 0/8 |

Expected global sum was `2080`. In direct/skew mode the four rank pairs
returned `206`, `414`, `624`, and `832`: exactly their own TP2-pair
contributions. Copy-in returned `2080` on every rank.

Instrumentation across 126,416 capture/registration events found zero
registration-count mismatches, peer IPC handle/offset mismatches, or
capture-to-registration drift. Strengthening the synchronization barrier also
failed 8/8, excluding the previous pointer-collision and simple visibility
hypotheses.

The #36088 workaround keeps custom AR enabled but stages graph inputs through
one stable pre-registered address per rank for `DSA + DP attention`. Full-model
direct registration failed 4/4; copy-in passed the focused gate and the 32+32
sequential/concurrent soak. Its measured decode throughput delta was -0.40%.
